src/NHibernate/Mapping/ForeignKey.cs

```diff
@@ -123,12 +123,22 @@ private void AlignColumns(Table referencedTable)
 				throw new FKUnmatchingColumnsException(sb.ToString());
 			}
 
-			using (var fkCols = ColumnIterator.GetEnumerator())
-			using (var pkCols = referencedTable.PrimaryKey.ColumnIterator.GetEnumerator())
+			AlignColumns(ColumnIterator, referencedTable.PrimaryKey.ColumnIterator);
+		}
+
+		internal static void AlignColumns(IEnumerable<Column> fk, IEnumerable<Column> pk)
+		{
+			using (var fkCols = fk.GetEnumerator())
+			using (var pkCols = pk.GetEnumerator())
 			{
 				while (fkCols.MoveNext() && pkCols.MoveNext())
 				{
-					fkCols.Current.Length = pkCols.Current.Length;
+					if (pkCols.Current.IsLengthDefined() || fkCols.Current.IsLengthDefined())
+						fkCols.Current.Length = pkCols.Current.Length;
+					if (pkCols.Current.IsPrecisionDefined() || fkCols.Current.IsPrecisionDefined())
```

Comment:
Previous column alignment logic between fk and pk was defining a length on fk even if it was not defined on pk, because those properties have default values (`255` for length). Now defining it only if it is defined on one side. I guess that if the case "defined on fk but not on pk" is possible, it will still fail, and it would be better to have a way to un-define it on fk. I have just made minimal changes here for at least ceasing specifying bad length on fk when it is actually not defined on pk.

And I have added alignement for scale and precision too.

Previous code was causing a bunch of tests to fail since max scale handling: the defined length was causing the column to convert `datetime` type to sql type through the method taking into account length/precision/scale, with default values for them (255 for length, 2 for scale). This was not having consequences when the capacity was only considered to be for size, since 255 was the out of range of datetime max scale, causing it to still use datetime default. But with the handling of scale as capacity instead, this logic was then causing fk to be generated with `datetime2(2)` sql type instead of `datetime2`.